CONTENT: two new blogs on package health (#116)
pyOpenSci/pyopensci.github.io: _posts/2022-11-07-what-makes-a-python-package-healthy.md

ecosystem. These packages tend to have smaller user bases, and smaller maintainer 
teams (or often are maintained by one volunteer person).

<figure>
    <a href="/images/python-packages/python-ecosystem-jake-vanderploss-pycon-2017.png"><img src="/images/python-packages/python-ecosystem-jake-vanderploss-pycon-2017.png" alt="Image used by Jake Vanderploss in the 2017 pycon conference that shows 
    the ecosystem of scientific python packages starting with foundational packages 
    and moving out to the wealth of smaller, domain specific packages."></a>
    <figcaption>Image used by Jake Vanderploss in the 2017 pycon conference that shows 
    the ecosystem of scientific python packages starting with foundational packages 
    and moving out to the wealth of smaller, domain specific packages. pyOpenSci
    focuses on the packages that are higher up in the ecosystem that often could 
    benefit from more support. 
    </figcaption>
</figure>

Our package maintainers:

* Often don't have the resources to build community
* Often are keen to build their user base and to contribute
to the broader scientific python ecosystem.   
</div>

> ### Three open source software healthy metric "buckets" 

Reviewer: the flow here loses me a little -- I think maybe you meant to remove this? since you repeat it below  

If this is meant to be a callback to previous posts, I would make that explicit.  

And/or insert a transition sentence to signal to your reader that you're not going to talk about these things yet: "Of course, many people have thought about metrics for package health, and this previous work came up in the conversation on twitter. Let me share those and acknowledge that work before I say more about this..."

    <figcaption>Here you can see what a Snyk report looks like for pandera, a package accepted into our ecosystem a few years ago. Pandera gets a very healthy report because it's heavily used among other metrics.
    </figcaption>
</figure>

#### Now let's look at [pyGMT package statistics on Snyk](https://snyk.io/advisor/python/pygmt)

{% include figure image_path="/images/pyopensci-packages/snyk-pygmt-python-package-metrics.png" alt="Image showing the metrics from the snyk website for the pyGMT package in our ecosystem" caption="In comparison to pandera, pyGMT gets a lower health score. I suspect this is due to lower community adoption and use. pyGMT is a much newer package. We'd argue however that pyGMT has a very healthy level of maintenance and even healthier package structure." %}

<figure>
    <a href="/images/pyopensci-packages/snyk-pygmt-python-package-metrics.png"><img src="/images/pyopensci-packages/snyk-pygmt-python-package-metrics.png" alt="Image showing the metrics from the snyk website for the pyGMT package in our ecosystem."></a>

>     <figcaption>In comparison to pandera, pyGMT gets a lower, but still good, health score. I suspect this is due to lower community adoption and use. pyGMT is a much newer package. We'd argue however that pyGMT has a very healthy level of maintenance and even healthier package structure.

Reviewer: I think this is a repeat of line 151 above? It shows up twice when I build locally too

Author: good catch. this is  a mistake! i was playing with two diff ways to embed images.

    </figcaption>
</figure>

> And of course the [scientific Python project has also been tracking the larger packages](https://devstats.scientific-python.org/):

Reviewer: did you mean for there to be an image here? Since the sentence ends with a colon

Might be nice too include an image from the link (if you weren't already planning on it)
e.g., the scipy one that lives at https://devstats.scientific-python.org/_images/scipy.png

Author: no this is just a link to see the live stats.

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">I suspect this is different from what a funder who is interested in sustainability or a corporation who is interested in their software supply chain would look for to define &quot;health&quot;</p>&mdash; Justin Kiggins (@neuromusic) <a href="https://twitter.com/neuromusic/status/1577744568203632640?ref_src=twsrc%5Etfw">October 5, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

Alas it is true that metrics designed for reporting that a funder requires
for a grant may differ metrics designed fom internal evaluation that 
informs program development. pyOpenSci has a lot to unpack there over the 
upcoming months!

### Three open source software healthy metric "buckets" 

> Based on all of the Twitter feedback (below), and what I *think* might be a 

Reviewer: this is a repeat from above, line 93 -- you meant to move it here?

I think it makes more sense here.

there are other package managers to consider too such as `Conda`. 

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Lots of thoughts on this... 😂<br>One aspect is definitely whether or not the package is published through PyPI, whether it follows standard packaging practices, has a test suite or well documented simple examples of how to use it, etc.</p>&mdash; Kenneth Hoste (@kehoste) <a href="https://twitter.com/kehoste/status/1578107988715671555?ref_src=twsrc%5Etfw">October 6, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

## Maintenance activity as a metric of health

The second topic that came up frequently on Twitter was the issue of maintenance. 

Jed Brown had some nice overarching insight here for things they look at that are 
indicators of both maintenance and bus factor (risk factor, mentioned below as a measure
of how many people / institutions support maintenance). More people and more institutions 
equals lower risk, fewer people and fewer institutions supporting the package equates
to a higher maintenance risk (or risk of the package becoming a sad orphan with 

> family to take care of it. 

Reviewer: ```suggestion
no family to take care of it. 
```
I think you mean "orphan with no family" not "orphan with family"

How many times have you tried to figure out what Python package you should use
 to process or download data, and you found 4 different packages on PyPI all in 
varying states of maintenance?

I've certainty been there. So has RenéKat it seems: 

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">I look to see if issues are being resolved. If it’s not being maintained I’m not going to waste my time installing it.</p>&mdash; RenéKat (@renekat14) <a href="https://twitter.com/renekat14/status/1577745550803451918?ref_src=twsrc%5Etfw">October 5, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

It's true. For a scientist (or anyone) it's a waste of time to install something
that won't be fixed as bugs arise. It's also not a good use of their time to have to dig into a package repository to see if it's being maintained or not. 

pyOpenSci does hope to help with this issue through a curated catalog of tools 
which will be developed over time. 

### But what constitutes maintenance?  

> How do we measure degree of maintenance? Number of issues being addressed And closed? Average commits each month, quarter or year?

Reviewer: ```suggestion
How do we measure degree of maintenance? Number of issues being addressed and closed? Average commits each month, quarter or year?
```
lowercase and

This could be a relative metrics too. Some package maintainers
may spend lots of time on issues or have too many to handle quickly as Melissa
points out replying to a comment about evaluating maintenance by looking 
at issues being closed:  

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Would that apply to large established projects such as NumPy? My guess is it wouldn&#39;t 😉</p>&mdash; Ax=13!!! (@melissawm) <a href="https://twitter.com/melissawm/status/1577776605220454403?ref_src=twsrc%5Etfw">October 5, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

> But alas I think there are ways around that. We can look at commits, pr's and such

Reviewer: ```suggestion
But alas I think there are ways around that. We can look at commits, pull request and such
```
suggest not abbreviating just to not confuse anyone

Yup

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">it is the equivalent of when a small bakery gets written up in the New York Times, has a huge influx of customers, and collapses under the weight of demand. I think it&#39;s an outcome we don&#39;t think about enough ahead of time</p>&mdash; Chris Holdgraf (@choldgraf) <a href="https://twitter.com/choldgraf/status/1578305233611915264?ref_src=twsrc%5Etfw">October 7, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

> > But I need atleast 5 (thousand) croissants, now. ANDDDD so does my friend. 

Reviewer: ```suggestion
> But I need at least 5 (thousand) croissants, now. ANDDDD so does my friend. 
```

Juan agrees that a steady stream of issues suggests adoption. Especially since 
opening issues on GitHub suggests that the users have some technical literacy. 

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">As others have said, it’s multidimensional, but this article argues that a steady stream of issues = a community of active and engaged users — often somewhat programming-literate since it’s GH. I find that argument compelling.<a href="https://t.co/X2vY2QxRfV">https://t.co/X2vY2QxRfV</a></p>&mdash; Juan Nunez-Iglesias (@jnuneziglesias) <a href="https://twitter.com/jnuneziglesias/status/1578290475017981953?ref_src=twsrc%5Etfw">October 7, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

## Metrics quantifying community around tools 

> I'd be remiss if i didn't atleast mention that some of the discussion steered 

Reviewer: ```suggestion
I'd be remiss if i didn't at least mention that some of the discussion steered 
```

towards community around tools. For instance, Evan brought up community governance
being a priority.

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Governance was another aspect I was going to suggest. The “benevolent dictator for life” model is… risky</p>&mdash; Evan (he/him) (@darth_mall) <a href="https://twitter.com/darth_mall/status/1577741937544609793?ref_src=twsrc%5Etfw">October 5, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

But the reality of our users was summarized well here by Tania. Most scientists 
developing tools are trying to simplify workflows with repeated code. Workflows
that others may be trying to develop to do the same thing. They aren't necessarily

> focused on community, atleast not yet. 

Reviewer: ```suggestion
focused on community, at least not yet. 
```
